### ISSUE TITLE
ux(sql): unhelpful assertion error message: `AssertionError: num_froms == 2` 

### ISSUE THREAD

**Issue by ogrisel** [ibis-project/ibis#4743]
Here is one way to trigger it:

```python
>>> import numpy as np
>>> import pandas as pd
>>> import ibis
>>> df = pd.DataFrame({"a": np.random.randn(10000)})
>>> t = ibis.memtable(df)
>>> t.mutate(chunk=ibis.row_number() // 100).group_by(ibis._.chunk).aggregate(b=t.a.mean()).execute()
Traceback (most recent call last):
  Input In [5] in <cell line: 5>
    t.mutate(chunk=ibis.row_number() // 100).group_by(ibis._.chunk).aggregate(b=t.a.mean()).execute()
  File ~/code/ibis/ibis/expr/types/core.py:274 in execute
    return self._find_backend(use_default=True).execute(
  File ~/code/ibis/ibis/backends/base/sql/__init__.py:247 in execute
    sql = query_ast.compile()
  File ~/code/ibis/ibis/backends/base/sql/compiler/base.py:39 in compile
    compiled_queries = [q.compile() for q in self.queries]
  File ~/code/ibis/ibis/backends/base/sql/compiler/base.py:39 in <listcomp>
    compiled_queries = [q.compile() for q in self.queries]
  File ~/code/ibis/ibis/backends/base/sql/alchemy/query_builder.py:186 in compile
    frag = step(frag)
  File ~/code/ibis/ibis/backends/base/sql/alchemy/query_builder.py:279 in _add_select
    assert num_froms == 1, f"num_froms == {num_froms:d}"
AssertionError: num_froms == 2
```

while using a deferred field reference would have worked:

```python
>>> t.mutate(chunk=ibis.row_number() // 100).group_by(ibis._.chunk).aggregate(b=ibis._.a.mean()).execute()
    chunk         b
0       0 -0.002539
1       1  0.211258
2       2 -0.034153
3       3  0.054345
4       4 -0.063897
..    ...       ...
95     83  0.077232
96     87 -0.018670
97     91 -0.153058
98     95  0.062328
99     99 -0.131931

[100 rows x 2 columns]
```

I am not sure if this is a bug or just a bad error message.

**Comment by cpcloud** [comment]
The assertion is correct but you shouldn't be hitting it. We need to catch the error earlier and give a more helpful and more usefully-typed message.

**Comment by ogrisel** [comment]
Out of curiosity, is there a way to chunk the rows without adding a new "chunk" field to the grouped table?

**Comment by cpcloud** [comment]
@ogrisel Can you explain a bit about what you're trying to do re chunking?

**Comment by cpcloud** [comment]
Generally speaking you'll probably have a bad time if you reference tables further down the tree than the current table. In some cases ibis tries to be smart and infer what you mean (in this case when you reference `t`), but fails to actually be smart and this is a case of that.

**Comment by ogrisel** [comment]
> @ogrisel Can you explain a bit about what you're trying to do re chunking?

It was an self-learning attempt at the first step of a proof of concept to see how we could turn scalar UDFs (as in PostgresSQL) into vector/table UDFs (as in duckdb) by rewriting a table expression to do the chunking ahead of time.

**Comment by NickCrews** [comment]
I just ran into this assertion error. Simplified example:

```python
import pandas as pd
import ibis
ibis.options.interactive = True

df = pd.DataFrame({"x": range(5)})
a = ibis.memtable(df)
b = a.mutate(a_x=a.x)
# The problem is that c refers to both a and b
c = b.mutate(b_x=b.x, a_x=a.x)
c
```
In my case I don't think I can use the `_` API, since I definitely want expression C to refer to both A and B.

I know that here in this trivial example I could do `c = b.mutate(b_x=b.x, a_x=b.a_x)` but in my real case this leads to ugly code and I would like to avoid it: I want to refer to intermediate expressions that I've calculated.

It seems like this example should work, right? Expression `a` is the only leaf node for expression `c`, so why can't ibis walk the dependencies and deal with this?

**Comment by cpcloud** [comment]
@NickCrews `b` contains all of `a`'s columns, so `_` should work for you here. Do you have an example of where that wouldn't work?

**Comment by NickCrews** [comment]
Yeah sorry for the non-representative example. `_` only works there because the `x` column is unchanged between all the tables. If we actually mutate it, then it is significant which of the previous tables we refer to:

```python
import pandas as pd
import ibis
ibis.options.interactive = True

def complex_calculation(t):
    return t.mutate(x = t.x + 1000)

df = pd.DataFrame({"x": range(5)})
inp = ibis.memtable(df)
computed = complex_calculation(inp)
result = inp.mutate(computed=computed.x)
result
```

This actually doesn't give the original assertion error, but the following stacktrace:

```python stacktrace
---------------------------------------------------------------------------
RelationError                             Traceback (most recent call last)
Cell In[13], line 11
      9 inp = ibis.memtable(df)
     10 computed = complex_calculation(inp)
---> 11 result = inp.mutate(computed=computed.x)
     12 result

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/expr/types/relations.py:607, in Table.mutate(self, exprs, **mutations)
    605 exprs.extend(ensure_expr(expr).name(name) for name, expr in mutations.items())
    606 mutation_exprs = an.get_mutation_exprs(exprs, self)
--> 607 return self.select(mutation_exprs)

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/expr/types/relations.py:701, in Table.select(self, *exprs, **named_exprs)
    696 if not exprs:
    697     raise com.IbisTypeError(
    698         "You must select at least one column for a valid projection"
    699     )
--> 701 op = an.Projector(self, exprs).get_result()
    703 return op.to_expr()

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/expr/analysis.py:535, in Projector.get_result(self)
    532     if fused_op is not None:
    533         return fused_op
--> 535 return ops.Selection(self.parent, self.clean_exprs)

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/common/grounds.py:23, in BaseMeta.__call__(cls, *args, **kwargs)
     22 def __call__(cls, *args, **kwargs) -> Base:
---> 23     return cls.__create__(*args, **kwargs)

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/common/grounds.py:100, in Annotable.__create__(cls, *args, **kwargs)
     96 @classmethod
     97 def __create__(cls, *args, **kwargs) -> Annotable:
     98     # construct the instance by passing the validated keyword arguments
     99     kwargs = cls.__signature__.validate(*args, **kwargs)
--> 100     return super().__create__(**kwargs)

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/common/grounds.py:31, in Base.__create__(cls, *args, **kwargs)
     29 @classmethod
     30 def __create__(cls, *args, **kwargs) -> Base:
---> 31     return type.__call__(cls, *args, **kwargs)

File ~/Library/Application Support/hatch/env/virtual/noatak-UM6-FHel/noatak/lib/python3.10/site-packages/ibis/expr/operations/relations.py:348, in Selection.__init__(self, table, selections, predicates, sort_keys, **kwargs)
    342 # print("----------")
    343 # print(table)
    344 # print(selections)
    345 # print(sort_keys)
    347 if not shares_all_roots(selections + sort_keys, table):
--> 348     raise com.RelationError(
    349         "Selection expressions don't fully originate from "
    350         "dependencies of the table expression."
    351     )
    353 for predicate in predicates:
    354     if isinstance(predicate, ops.Literal):

RelationError: Selection expressions don't fully originate from dependencies of the table expression.
```

As I've been going around refactoring my code, I keep on swapping between getting either one of these two errors. I figured they were stemming from a similar problem, but sorry if this is unrelated. I can open a separate issue, and try to find another example that actually produces the original assertion error.

**Comment by cpcloud** [comment]
@NickCrews These issues are related, and I can understand that it's frustrating to have to workaround something that you expect should just work.

I would like to understand more about whether there's **something** we can do to make your current situation less annoying.

Is there an actual code sample you can show?

**Comment by cpcloud** [comment]
After #5213, this no longer errors and produces warning from SQLAlchemy indicating the exact problem with referencing tables in arbitrary places: it's easy to generate a cross join, because (nearly) every operation creates a new relation and it's very difficult to correctly guess 1) what the user intended (there are often multiple ways to interpret child references) and 2) ensure that the sql is correct for the intention.

I'm going to close this out because the original issue is fixed (the unhelpful error message). We can open another issue for the cross join if it's a problem in practice.
